# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: servo/components/selectors/builder.rs
# Commit: 73ee44ffb320
# Full Hash: 73ee44ffb32056a47a020486d3f8d1ba5dfe381f
# Author: Emilio Cobos Álvarez <emilio@crisal.io>
# Date: 2023-07-12 21:49:07
# Regressor Bug: 1842221
# File Overlap Count: 1
# Description:
#   Bug 1842221 - Simplify ThinArc and friends. r=boris
#   
#   ThinArc is way more complex than it needs to be, and miri complains
#   about various things when using the selectors crate with it.
#   
# ==============================================================================

diff -r c32acbd14a56 -r 73ee44ffb320 servo/components/selectors/builder.rs
--- a/servo/components/selectors/builder.rs	Wed Jul 12 07:03:02 2023 +0000
+++ b/servo/components/selectors/builder.rs	Wed Jul 12 07:17:25 2023 +0000
@@ -19,7 +19,7 @@
 
 use crate::parser::{Combinator, Component, RelativeSelector, Selector, SelectorImpl};
 use crate::sink::Push;
-use servo_arc::{Arc, HeaderWithLength, ThinArc};
+use servo_arc::{Arc, ThinArc};
 use smallvec::{self, SmallVec};
 use std::cmp;
 use std::iter;
@@ -105,32 +105,22 @@
         &mut self,
         spec: SpecificityAndFlags,
     ) -> ThinArc<SpecificityAndFlags, Component<Impl>> {
-        // First, compute the total number of Components we'll need to allocate
-        // space for.
-        let full_len = self.simple_selectors.len() + self.combinators.len();
-
-        // Create the header.
-        let header = HeaderWithLength::new(spec, full_len);
-
         // Create the Arc using an iterator that drains our buffers.
-
-        // Use a raw pointer to be able to call set_len despite "borrowing" the slice.
-        // This is similar to SmallVec::drain, but we use a slice here because
-        // we’re gonna traverse it non-linearly.
-        let raw_simple_selectors: *const [Component<Impl>] = &*self.simple_selectors;
-        unsafe {
-            // Panic-safety: if SelectorBuilderIter is not iterated to the end,
-            // some simple selectors will safely leak.
-            self.simple_selectors.set_len(0)
-        }
-        let (rest, current) = split_from_end(unsafe { &*raw_simple_selectors }, self.current_len);
+        // Panic-safety: if SelectorBuilderIter is not iterated to the end, some simple selectors
+        // will safely leak.
+        let raw_simple_selectors = unsafe {
+            let simple_selectors_len = self.simple_selectors.len();
+            self.simple_selectors.set_len(0);
+            std::slice::from_raw_parts(self.simple_selectors.as_ptr(), simple_selectors_len)
+        };
+        let (rest, current) = split_from_end(raw_simple_selectors, self.current_len);
         let iter = SelectorBuilderIter {
             current_simple_selectors: current.iter(),
             rest_of_simple_selectors: rest,
             combinators: self.combinators.drain(..).rev(),
         };
 
-        Arc::into_thin(Arc::from_header_and_iter(header, iter))
+        Arc::from_header_and_iter(spec, iter)
     }
 }
 